### PR TITLE
Add SslStream test for zero-byte reads

### LINKED ISSUE #37122: Support 0 byte reads with SSLStream
Bufferless reads are a technique used in a couple of places in ASP.NET Core to avoid renting memory from the memory pool if there's no data available on the socket; this helps when trying to reduce the memory usage for idle connections for applications that are trying to handle a large number of concurrent connections (like websocket based applications).
- Sockets - https://github.com/dotnet/aspnetcore/blob/278715734e957c8cfb3ee720510a79e8db1495b4/src/Servers/Kestrel/Transport.Sockets/src/Internal/SocketConnection.cs#L195 - Aside, it would also be great if we could make this cheaper 😄 
- SignalR - https://github.com/dotnet/aspnetcore/blob/278715734e957c8cfb3ee720510a79e8db1495b4/src/SignalR/common/Http.Connections/src/Internal/Transports/WebSocketsServerTransport.cs#L147-L153

0 byte reads are supported by Sockets today and we take advantage of that to support the above scenario. We'd like to do the same for HTTPs based connections, so it would need to also work for SSLStream.

The typical layering looks like this for a WebSocket implemention (roughly):

```
WS.ReadAync(Buffer) 
    -> US.ReadAync(WSHeaderBuffer)
    -> US.ReadAync(Buffer)
        -> SSlStream.ReadAync(TlsFrameBuffer)
        -> SSlStream.ReadAync(Buffer) 
            -> NetworkStream.ReadAync(Buffer)
                -> Socket.ReadAsync(Buffer)
```

PS: I haven't tested it but if it just works then we should add test coverage so it doesn't regress.

cc: @stephentoub @wfurt

### PR DESCRIPTION
**Pull request by stephentoub**
We have higher-level tests for libraries like WebSockets that validate the behavior of zero-byte reads, but we don't currently have one in System.Net.Security's tests.  Adding one.

Closes https://github.com/dotnet/runtime/issues/37122
cc: @wfurt

### CODE REVIEW

File: src/libraries/System.Net.Security/tests/FunctionalTests/SslStreamStreamToStreamTest.cs [hunk 1]
```diff
@@ -259,6 +259,38 @@ public async Task SslStream_StreamToStream_Successive_ClientWrite_WithZeroBytes_
             }
         }
 
+        [Fact]
+        public async Task SslStream_StreamToStream_ZeroByteRead_SucceedsWhenDataAvailable()
+        {
+            using var listener = new Socket(AddressFamily.InterNetwork, SocketType.Stream, ProtocolType.Tcp);
```
Comment (wfurt) [inline src/libraries/System.Net.Security/tests/FunctionalTests/SslStreamStreamToStreamTest.cs:265]: You can use TestHelper.GetConnectedTcpStreams()

**Review by wfurt — APPROVED** [review]
LGTM. There is helper function you can use to simplify TCP connect.

File: src/libraries/System.Net.Security/tests/FunctionalTests/SslStreamStreamToStreamTest.cs [hunk 2]
```diff
@@ -259,6 +259,38 @@ public async Task SslStream_StreamToStream_Successive_ClientWrite_WithZeroBytes_
...
+                Assert.False(zeroByteRead.IsCompleted);
+
+                await serverSslStream.WriteAsync(Encoding.UTF8.GetBytes("hello"));
+                Assert.Equal(0, await zeroByteRead);
+
+                var readBytes = new byte[5];
+                int count = 0;
+                while (count < readBytes.Length)
+                {
+                    count += await clientSslStream.ReadAsync(readBytes.AsMemory(count));
```
Comment (halter73) [inline src/libraries/System.Net.Security/tests/FunctionalTests/SslStreamStreamToStreamTest.cs:288]: Should this verify the return value is non-zero?

**Review by halter73 — APPROVED** [review]

File: src/libraries/System.Net.Security/tests/FunctionalTests/SslStreamStreamToStreamTest.cs (hunk 1 above)
Comment (stephentoub) [reply src/libraries/System.Net.Security/tests/FunctionalTests/SslStreamStreamToStreamTest.cs]: Thanks.  Changed to use that.

File: src/libraries/System.Net.Security/tests/FunctionalTests/SslStreamStreamToStreamTest.cs (hunk 2 above)
Comment (stephentoub) [reply src/libraries/System.Net.Security/tests/FunctionalTests/SslStreamStreamToStreamTest.cs]: Sure, added an assert.

**Review by halter73 — APPROVED** [review]

**Review by davidsh — APPROVED** [review]

**Pull request merged**